### PR TITLE
feat: Added GetDetailedFeatureDecisionUnsafe

### PR DESCRIPTION
**Pull request by msohailhussain**
### Summary
Temporary changes in go-sdk for adding a new method GetDetailedFeatureDecisionUnsafe
This is added to support some agent features. Which will be removed later on.

### Testplan
- All FSC should be passed.

### CODE REVIEW

File: pkg/client/client.go [hunk 1]
```diff
@@ -467,6 +447,66 @@ func (o *OptimizelyClient) GetAllFeatureVariablesWithDecision(featureKey string,
...
+	for _, v := range feature.VariableMap {
+		val := v.DefaultValue
+
+		if decisionInfo.Enabled {
+			if variable, ok := featureDecision.Variation.Variables[v.ID]; ok {
+				val = variable.Value
+			}
+		}
+
+		out, typedError := o.getTypedValue(val, v)
```
Comment (mjc1283) [inline pkg/client/client.go:493]: Nit: `out` isn't an ideal variable name IMO. What does it mean?

File: pkg/client/client.go [hunk 2]
```diff
@@ -467,6 +447,66 @@ func (o *OptimizelyClient) GetAllFeatureVariablesWithDecision(featureKey string,
...
+
+	errs := new(multierror.Error)
+
+	for _, v := range feature.VariableMap {
+		val := v.DefaultValue
+
+		if decisionInfo.Enabled {
+			if variable, ok := featureDecision.Variation.Variables[v.ID]; ok {
+				val = variable.Value
+			}
```
Comment (mjc1283) [inline pkg/client/client.go:490]: If `ok` is false, let's log an error.

File: pkg/client/client.go [hunk 3]
```diff
@@ -467,6 +447,66 @@ func (o *OptimizelyClient) GetAllFeatureVariablesWithDecision(featureKey string,
...
+		decisionInfo.VariableMap[v.Key] = out
+	}
+
+	if o.notificationCenter != nil {
+		decisionNotification := decision.FeatureNotificationWithVariables(featureKey, &featureDecision, &userContext,
+			map[string]interface{}{"variableValues": decisionInfo.VariableMap})
+		decisionNotification.Type = notification.AllFeatureVariables
+
+		if err = o.notificationCenter.Send(notification.Decision, *decisionNotification); err != nil {
+			o.logger.Warning("Problem with sending notification")
```
Comment (mjc1283) [inline pkg/client/client.go:504]: Let's log `err` here so the user can see what the problem was.

File: pkg/client/client.go [hunk 4]
```diff
@@ -685,6 +725,26 @@ func (o *OptimizelyClient) RemoveOnTrack(id int) error {
 	return nil
 }
 
+func (o *OptimizelyClient) getTypedValue(value string, variable entities.Variable) (out interface{}, err error) {
+	out = value
+	switch varType := variable.Type; varType {
```
Comment (mjc1283) [inline pkg/client/client.go:730]: Why do we need to use a new variable `varType`? Does this not work?
```suggestion
	switch variable.Type {
```

File: pkg/client/client.go [hunk 5]
```diff
@@ -467,6 +447,66 @@ func (o *OptimizelyClient) GetAllFeatureVariablesWithDecision(featureKey string,
...
+	decisionContext, featureDecision, err := o.getFeatureDecision(featureKey, "", userContext)
+	if err != nil {
+		o.logger.Error("Optimizely SDK tracking error", err)
+		return decisionInfo, err
+	}
+
+	if featureDecision.Variation != nil {
+		decisionInfo.Enabled = featureDecision.Variation.FeatureEnabled
+		decisionInfo.VariationKey = featureDecision.Variation.Key
+		decisionInfo.ExperimentKey = featureDecision.Experiment.Key
```
Comment (mjc1283) [inline pkg/client/client.go:466]: I think we should only assign these when `featureDecision.Source == decision.FeatureTest`. Keys from a rollout experiment should not be exposed.

File: pkg/client/client.go [hunk 6]
```diff
@@ -685,6 +725,26 @@ func (o *OptimizelyClient) RemoveOnTrack(id int) error {
 	return nil
 }
 
+func (o *OptimizelyClient) getTypedValue(value string, variable entities.Variable) (out interface{}, err error) {
```
Comment (pawels-optimizely) [inline pkg/client/client.go:728]: do we need to pass the whole variable ? Seems like only type will be enough.

File: pkg/client/client_test.go [hunk 7]
```diff
@@ -1679,6 +1679,294 @@ func TestGetAllFeatureVariablesWithDecisionWithoutFeature(t *testing.T) {
...
+	testFeatureKey := "test_feature_key"
+	testUserContext := entities.UserContext{ID: "test_user_1"}
+
+	type variable struct {
+		key        string
+		defaultVal string
+		varVal     string
+		varType    entities.VariableType
+		expected   interface{}
+	}
```
Comment (mjc1283) [inline pkg/client/client_test.go:1789]: The same struct type is declared in the previous test. I suggest declaring it only once inside this file.

File: pkg/client/client_test.go [hunk 8]
```diff
@@ -1679,6 +1679,294 @@ func TestGetAllFeatureVariablesWithDecisionWithoutFeature(t *testing.T) {
...
+
+		variableMap[id] = entities.Variable{
+			DefaultValue: v.defaultVal,
+			ID:           id,
+			Key:          v.key,
+			Type:         v.varType,
+		}
+
+		mockConfig.On("GetVariableByKey", testFeatureKey, v.key).Return(v.varVal, nil)
+	}
```
Comment (mjc1283) [inline pkg/client/client_test.go:1722]: This logic is repeated in several tests, please factor this out into a helper function.

**Pull request merged**